### DISCUSSION TITLE
Is it possible to configure renovate to not use the sidecar docker image now?

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#19740] [Request Help] [answered]
I have forked renovate and have added some extra code in renovate which for certain files end up running dotnet build for some files and includes the output as part of the renovate PRs.
I used to be able to do:
```
const execOptions: ExecOptions = {
        docker: {
          image: 'dotnet',
        },
        extraEnv: { NUGET_PACKAGES: upath.join(nugetCacheDir, 'packages') },
      };
```
but since #19545 that is no longer possible since `image` was removed.
I have tried running with this instead but it ends up giving other errors
``` toolConstraints: [
          { toolName: 'dotnet', constraint: config.constraints?.dotnet },
        ],
```
I cannot find a way in the code to still allow it to run it in the `renovate/dotnet` image instead of the sidecar image.
Is there something I'm missing?

**Comment by rarkins** [comment, marked as answer]
Those images are now deprecated so you wouldn't want to go down that path anyway

**Reply by HenrikPoulsen** [reply]
Thanks good to know.
Then I guess I need to somehow resolve this error or maintain my own docker images.
The issue I'm seeing is the following.
exec tries to run the following in the sidecar image:
```
install-tool dotnet 7.0.101 && dotnet build -c Release -o <pathForBuildOutput> <pathToProjectInRepo> && dotnet <pathToGeneratedDll>
```
it installs dotnet 7.0.101 as intended and builds the project successfully but then the building fails with the following when it tries to run it:
```
You must install or update .NET to run this application.

App: <pathToDll>
Architecture: x64
Framework: 'Microsoft.NETCore.App', version '6.0.0' (x64)
.NET location: /opt/buildpack/tools/dotnet/

The following frameworks were found:
  7.0.1 at [/opt/buildpack/tools/dotnet/shared/Microsoft.NETCore.App]

Learn about framework resolution:
[https://aka.ms/dotnet/app-launch-failed](https://aka.ms/dotnet/app-launch-failed)

To install missing framework, download:
https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=6.0.0&arch=x64&rid=ubuntu.20.04-x64

```
So it's not as simple as just installing the latest version, since obviously the project can reference older framework versions as is the case here.
It also does make things a lot slower (we run this many times per run in some repos) to rely on `install-tool` rather than just spinning up a container that already has everything installed.
So it does look like I will need to maintain my own dotnet image.

So am I correct that there is no intended way currently of overriding the sidecar image anymore and I will need to basically revert some of the changes from #19545 to be able to control the image used again?

**Reply by HenrikPoulsen** [reply]
Update: Regarding the install version. I was wrong about the project needing dotnet 7, it has indeed specified dotnet 6 in the csproj file, but I am not supplying that. So I would need to resolve that as well to ensure the right version is used.

That leaves the slowness. My docker command invocation knowledge is not that high. Is the container reused in any way between the execs or is the install-tool indeed executed freshly between each exec command?

**Reply by rarkins** [reply]
install-tool is executed freshly, although there can be some caching of those tools so that they don't need to be downloaded again.

**Reply by viceice** [reply]
Use `dockerImagePrefix`[^1] to create your own sidecar image

[^1]: https://docs.renovatebot.com/self-hosted-configuration/#dockerimageprefix

**Reply by HenrikPoulsen** [reply]
> although there can be some caching of those tools so that they don't need to be downloaded again.

Is that something I need to do to have enabled? Out of the box it does not seem to do that and is just quite slow whenever it needs to run something for dotnet

> Use dockerImagePrefix[1](https://github.com/renovatebot/renovate/discussions/19740#user-content-fn-1-b48871d74f4716f9766c70eb260ed862) to create your own sidecar image

Thanks, would have preferred to have more control over it. Since doing that would require me making a monolith image that have all past and current versions of all the tool versions installed. Or can it be constrained somehow to do `myprefix-renovate/sidecar:dotnet-6` etc?

**Reply by HenrikPoulsen** [reply]
Some metrics.
It does seem like the caching does work, so the first time the command I mentioned before runs it takes ~33seconds in total.
Second time and onwards (it runs it once for each dependency update PR) it takes ~13 seconds.

Comparing this to before the image support was dropped the first run would take ~10 seconds
and subsequent runs ~3 seconds.
The discepancy there is that it does incremental builds on the reruns so it saves a lot of time there on the nuget restore and building.

So for our larger monorepo it is going to add a lot of extra time with this ~4x execution time increase for that particular command.

**Reply by viceice** [reply]
you probably should use the full image, it has mostly latest version of all tools preinstalled and installs missing tools on the fly

**Reply by HenrikPoulsen** [reply]
Thanks I'll take a look. I'll mark this as answered
